SOLR-16478 : Further changes to avoid PRS entry writes from overseer

patsonluk opened the pull request:
https://issues.apache.org/jira/browse/SOLR-16478

# Description

For further improvements based on https://issues.apache.org/jira/browse/SOLR-16414, there are places that we should remove PRS entry operations from overseer:

1. In `ZkController#unregister`, the PRS hosting node should remove the PRS entries from  ZK, as overseer no longer does so as in the change of https://github.com/apache/solr/pull/1019/files#diff-b08497f972c47127009b59dc49de322698312c08ee53684d6e916db256db16a2L149
2. We should as well remove the PRS entries ops from Overseer in `SliceMutator#addReplica` https://github.com/apache/solr/blob/c96c78132d77974adaec62ffaaf4b819f76f49a9/solr/core/src/java/org/apache/solr/cloud/overseer/SliceMutator.java#L112-L123

# Solution

For `ZkController#unregister`, add operations to remove the PRS entries
For  `SliceMutator#addReplica` , it should no longer perform operation to add PRS entries

# Tests

For `ZkController#unregister`, enhanced the existing test case `PerReplicaStatesIntegrationTest#testPerReplicaStateCollection`, to test on replica removal as well. Before the fix, the PRS state fetched from `PerReplicaStatesFetcher` would show that the replica was NOT removed. But after the fix, it show the correct PRS state count.

For `SliceMutator#addReplica`, it is a bit tricky to verify. We added a new add replcia call and ZK child node version check in `PerReplicaStatesIntegrationTest#testZkNodeVersions`, though the child node version is the same before and after the fix. As before the fix, the `AddReplicaCmd` which calls `SliceMutator#addReplica` would add the PRS entries (which is incorrect), but even if we removed the PRS op from `SliceMutator#addReplica`, the state.json insertion would still "TOUCH" the PRS entries, hence ended up with the same child version increment. But anyway, it is still good with the new test case to verify the child version number.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [ ] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [x] I have added tests for my changes.
- [ ] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

noblepaul commented:
@patsonluk the changes are correct and we need to update the `testZkNodeVersions()` reflect it. 

I think this should go into `9.1`

patsonluk commented:
> @patsonluk the changes are correct and we need to update the `testZkNodeVersions()` reflect it.
> 
> I think this should go into `9.1`

Thanks @noblepaul ! I have added another test op (deleteReplica) in `testZkNodeVersions` and adjusted the numbers :)

As for going into `9.1`, I assume that would happen after this being merged to `master`, and you can help us with that? 🙏🏼
